STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/_az_ulib_dm_blob.c

```diff
@@ -174,70 +174,111 @@ static az_result copy_blob_to_flash(NXD_ADDRESS* ip, CHAR* resource, CHAR* host,
-                                  NX_WEB_HTTP_METHOD_GET, /* GET, PUT, DELETE, POST, HEAD */
-                                  resource, /* "resource" (usually includes auth headers)*/
-                                  host, /* "host" */
-                                  0, /* Used by PUT and POST */
-                                  NX_FALSE, /* If true, input_size is ignored. */
-                                  NX_NULL, /* "name" */
-                                  NX_NULL, /* "password" */
+        // add user agent
+        if((nx_status = nx_web_http_client_request_header_add(&http_client, 
+                                  USER_AGENT_NAME, 
+                                  sizeof(USER_AGENT_NAME) - 1, 
+                                  USER_AGENT_VALUE, 
+                                  sizeof(USER_AGENT_VALUE) - 1, 
                                   DCF_WAIT_TIME)) == NX_SUCCESS)
         {
-          if((nx_status = nx_web_http_client_request_header_add(&http_client, 
-                                    USER_AGENT_NAME, 
-                                    sizeof(USER_AGENT_NAME) - 1, 
-                                    USER_AGENT_VALUE, 
-                                    sizeof(USER_AGENT_VALUE) - 1, 
-                                    DCF_WAIT_TIME)) == NX_SUCCESS)
-          {
-            nx_status = nx_web_http_client_request_send(&http_client, DCF_WAIT_TIME);
-            // Download destination pointer, increase after every packet is stored
-            uint32_t total_downloaded_size = 0;
-            while(nx_status == NX_SUCCESS)
+          // send request
+          if((nx_status = nx_web_http_client_request_send(&http_client, 
+                                                          DCF_WAIT_TIME)) == NX_SUCCESS)
             {
+              // grab first chunk
               nx_status = nx_web_http_client_response_body_get(&http_client, &packet_ptr, 500);
-              if((nx_status == NX_SUCCESS) || (nx_status == NX_WEB_HTTP_GET_DONE))
-              {
-                az_span data = az_span_create(packet_ptr->nx_packet_prepend_ptr, 
-                                              packet_ptr->nx_packet_length);
 
-                // calculate destination pointer
-                uint8_t* dest_ptr = (uint8_t*)(address) + total_downloaded_size;
+              // save pointer to first chunk
+              az_span data = az_span_create(packet_ptr->nx_packet_prepend_ptr, 
+                                                  packet_ptr->nx_packet_length);
+
+              // calculate destination pointer
+              uint32_t total_downloaded_size = 0;
+              uint8_t* dest_ptr = (uint8_t*)(address) + total_downloaded_size;
+
+              // grab package size and round up to nearest 2KB (0x0800)
+              unsigned int package_size = (unsigned int)http_client.nx_web_http_client_total_receive_bytes;
+              package_size += 0x0800 - package_size % 0x0800;
```

Comment:
>package_size += 0x0800 - package_size % 0x0800; [](http://example.com/codeflow?start=14&length=47)

Looks a little expensive for me, consider something like
```
if((package_size & 0x07FF) != 0x000)
{
  package_size = (package_size & 0xFFFFF800) + 0x0800;
}
```

Reply:
oh wow yes much more elegant.